src/v/storage/segment.cc

```diff
@@ -340,7 +340,7 @@ ss::future<> segment::do_flush() {
         // outstanding flushes once the one executed later in terms of offset
         // finishes we guarantee that all previous flushes finished.
         _tracker.committed_offset = std::max(o, _tracker.committed_offset);
-        _tracker.stable_offset = _tracker.committed_offset;
+        _tracker.stable_offset = std::max(o, _tracker.stable_offset);
```

Comment:
Trying to connect the dots here: should `o` here be exactly the same as what was advanced to during the appender flush?

Reply:
Also did you have a specific sequence in mind that leads to this not being non-monotonic previously? That would really help understand this change (this change looks much more intuitive, but just want to know what went wrong before)

Reply:
> Trying to connect the dots here: should o here be exactly the same as what was advanced to during the appender flush?

Yes.

> Also did you have a specific sequence in mind that leads to this not being non-monotonic previously? That would really help understand this change (this change looks much more intuitive, but just want to know what went wrong before)

The updated test fails without this change.

A issues a flush, stores the dirty offset in `o`
B does 10 additional writes , issues a flush, stores the dirty offset in `o`, actually flushes, updates stable_offset to N+10
A finally does the actual appender::flush, moves the stable_offset back to N overwriting the value set by B